### PR TITLE
[Merged by Bors] - feat(ModelTheory): Ax-Grothendieck

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---
- [x] depends on: #6453 
- [x] depends on: #6472 
- [x] depends on: #6518 
- [x] depends on: #6566  
- [x] depends on: #6613 
- [x] depends on: #6617 
- [x] depends on: #6558
- [x] depends on: #6635 
- [x] depends on: #6649 
- [x] depends on: #6650
- [x] depends on: #6651
- [x] depends on: #6654 
- [x] depends on: #6656 
- [x] depends on: #6658 
- [x] depends on: #7185 
- [x] depends on: #7186 
- [x] depends on: #7187 
- [x] depends on: #7188 
- [x] depends on: #16031 
- [x] depends on: #16030
- [x] depends on: #16019
- [x] depends on: #16035 
- [x] depends on: #16037 
- [x] depends on: #16503 
- [x] depends on: #16504 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib.lean [hunk 1]
```diff
@@ -3550,10 +3550,11 @@ import Mathlib.ModelTheory.Algebra.Field.Basic
...
 import Mathlib.ModelTheory.Algebra.Field.IsAlgClosed
 import Mathlib.ModelTheory.Algebra.Ring.Basic
+import Mathlib.ModelTheory.Algebra.Ring.Definability
 import Mathlib.ModelTheory.Algebra.Ring.FreeCommRing
 import Mathlib.ModelTheory.Basic
+import Mathlib.ModelTheory.Basic
 import Mathlib.ModelTheory.Bundled
-import Mathlib.ModelTheory.Complexity
 import Mathlib.ModelTheory.Definability
 import Mathlib.ModelTheory.DirectLimit
```
[the remark this answers is not in the document]
Comment (ChrisHughes24) [reply Mathlib.lean:3559]: bot fix style

File: Mathlib/FieldTheory/AxGrothendieck.lean [hunk 2]
```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
...
+also surjective on `S`.
+* `ax_grothendieck_univ`: Any injective polynomial map `K^n → K^n` is also surjective if `K` is an
+algberaically closed field.
+* `ax_grothendieck_of_definable`: Any injective polynomial map `S → S` is also surjective on `S` if
+`K` is an algebraically closed field and `S` is a definable subset of `K^n`.
+* `ax_grothendieck_of_locally_finite`: any injective polynomial map `R^n → R^n` is also surjective
+whenever `R` is an algebraic extension of a finite field.
+
+## References
+The first order theory of algebraically closed fields, along with the Lefschetz Principle and
```
Comment (jcommelin) [inline Mathlib/FieldTheory/AxGrothendieck.lean:32]:
```suggestion

The first order theory of algebraically closed fields, along with the Lefschetz Principle and
```

File: Mathlib/FieldTheory/AxGrothendieck.lean [hunk 3]
```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
...
-then any injective polynomial map `R^n → R^n` is also surjective.
+This file proves that if `K` is an algebraically closed field,
+then any injective polynomial map `K^n → K^n` is also surjective.
 
-This proof is required for the true Ax-Grothendieck theorem, which proves the same result
-for any algebraically closed field of characteristic zero.
+## Main results
 
-## TODO
+* `ax_grothendieck_zeroLocus`: If `K` is algebraically closed and `S : Set (ι → K)`
```
Comment (jcommelin) [inline Mathlib/FieldTheory/AxGrothendieck.lean:21]: I guess it is understood, but maybe mention that `ι` is finite.

File: Mathlib/FieldTheory/AxGrothendieck.lean [hunk 4]
```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
...
+
+end FirstOrder
+
+open FirstOrder Language Field Ring MvPolynomial
+
+variable {K ι : Type*} [Field K] [IsAlgClosed K] [Finite ι]
+
+/-- A slight generalization of the **Ax-Grothendieck** theorem
+
+If `K` is an algebraically closed field, and `S` is a definable subset of `ι → K`, then any
```
Comment (jcommelin) [inline Mathlib/FieldTheory/AxGrothendieck.lean:199]: Here as well: I guess it is understood, but maybe mention that `ι` is finite.

**Review by jcommelin — COMMENTED** [review]
Nice!

File: Mathlib/FieldTheory/AxGrothendieck.lean (hunk 4 above)
Comment (ChrisHughes24) [reply Mathlib/FieldTheory/AxGrothendieck.lean]: done

File: Mathlib/FieldTheory/AxGrothendieck.lean [hunk 5]
```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
...
+    let S := zeroLocus I
+    S.MapsTo (fun v i => eval v (p i)) S →
+    S.InjOn (fun v i => eval v (p i)) →
+    S.SurjOn (fun v i => eval v (p i)) S := by
+  letI := compatibleRingOfRing K
+  intro S
+  have i_fg : I.FG := IsNoetherian.noetherian I
+  rcases i_fg with ⟨s, rfl⟩
+  exact ax_grothendieck_of_definable
+    S (mvPolynomial_zeroLocus_definable s) p
```
Comment (jcommelin) [inline Mathlib/FieldTheory/AxGrothendieck.lean:240]:
```suggestion
  obtain ⟨s, rfl⟩ : I.FG := IsNoetherian.noetherian I
  exact ax_grothendieck_of_definable S (mvPolynomial_zeroLocus_definable s) p
```

File: Mathlib/FieldTheory/AxGrothendieck.lean [hunk 6]
```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
...
+
+/-- A special case of the **Ax-Grothendieck** theorem
+
+Any injective polynomial map `K^n → K^n` is also surjective if `K` is an
+algberaically closed field. -/
+theorem ax_grothendieck_univ (p : ι → MvPolynomial ι K) :
+    (fun v i => eval v (p i)).Injective →
+    (fun v i => eval v (p i)).Surjective := by
+  simpa [Set.injective_iff_injOn_univ,
+         Set.surjective_iff_surjOn_univ] using
```
Comment (jcommelin) [inline Mathlib/FieldTheory/AxGrothendieck.lean:250]:
```suggestion
  simpa [Set.injective_iff_injOn_univ, Set.surjective_iff_surjOn_univ] using
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**